File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
  // this is naive recursion + caching approach, see editorail for efficient DP approach
    int length_from(int[] nums,int pos,int[] cache){
      if(pos >= nums.length) return 0;
      if(pos == nums.length-1){cache[pos]=1;return 1;}
      if(cache[pos] == -1){
        int max=0;
        for(int i=pos+1;i<=nums.length-1;i++){
          if (nums[pos]<nums[i]){
            if (max < length_from(nums,i,cache)) max= length_from(nums,i,cache);
          }
        }
        cache[pos] = 1+max; 
      }
      return cache[pos];
    }
    public int lengthOfLIS(int[] nums) {
        // if(nums.length == 1) return 1;int max = Integer.MIN_VALUE;
        // int[] cache = new int[nums.length]; Arrays.fill(cache,-1);
        // for(int i=nums.length-1;i>=0;i--){
        //   if (max < length_from(nums,i,cache)) max = length_from(nums,i,cache);
        // }
        // return max;

        // here you can see the iterative nature of the problem, it is easy to solve with bottom up dp approach
        // int dp[] = new int[nums.length];Arrays.fill(dp,1);
        // for(int i=1; i<nums.length; ++i){
        //   for(int j=0; j<i; ++j){
        //     if(nums[j] < nums[i])
        //       dp[i] = Math.max(dp[i],dp[j]+1);
        //   }
        // }
        // int ans= Integer.MIN_VALUE;
        // for(int i: dp) ans = Math.max(ans,i);
        // return ans;

         // there is another way to do this, by continusly building the sequence (time : O(n*logn))
    ArrayList<Integer> seq = new ArrayList<>();
    seq.add(nums[0]); //lets start by including first number
    for(int i=1; i<nums.length; ++i){
      int num = nums[i];
      if(num > seq.get(seq.size()-1)) { // check last number in seq, if its lower than num, include in seq
          seq.add(num);
      }else{
         //if that num is less than last num is seq, there is a chance that this can be used to build even bigger sequence, so we need to include in sequence but in correct position without disturbing order of previous seq (so we will us binary search to find suitable position)
         int pos = Collections.binarySearch(seq, num);
         if(pos < 0) pos = -(pos)-1;
         seq.set(pos, num);
      }
    }
    return seq.size();
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
import bisect

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:

        # first lets try with naive recursion + memo approach

        '''
        
        @functools.cache
        def recurse(index : int) -> int : 

          # base case is 
          if (index < 0):
            return 0

          count = 1
        
          for i in range(index):
            if nums[index] > nums[i] :
              count = max(count, recurse(i)+1)

          return count

        
        ans =0
        
        for i in range(len(nums)):
          ans = max(ans, recurse(i))

        return ans
          
        '''
        # lets try with bottom up dp (it is easy to do, beacuse of recursive nature of problem) see : https://www.youtube.com/watch?v=cjWnW0hdF1Y

        '''

        #as usaual first we need to take the dp List
        dp = [1] * len(nums)  # since, every number is itself counted as 1

        for i in range(len(nums)-1, -1, -1):   # we starting from last number in given array

          for j in range(i+1, len(nums)):

            if nums[i] < nums[j]:
                
              dp[i] = max( dp[i] , 1 + dp[j])

        return max(dp)

        '''

        # lets deal with follow up question to solve this in O(n* logn) as above methods runtime is O(n^2)

        # trick is that, we just travel through the list, for example, list is [8, 1, 6, 2, 3, 10]. first we will take, after that 1 is there, so replace 8 with 1, next 6 is there which is >1, so keep it [1,6], next 2 is there which is less that 6, {which might be useful to construct bigger sequence}, so replace with 6. But we need to think about where actually we need to place this new low element in out result list. {answer is, using binary search to find perfect position n our newly forming list, as this list is alreay sorted in ascending order}

        ans = [nums[0]]
        nums1 = nums[1:]
        #lets travel through given list
        for num in nums1:
          if num > ans[len(ans)-1]:  #if num is big, lets keep at end of ans
            ans.append(num)
          else:
            pos = bisect.bisect_left(ans, num)  #if num is small, its has potential, so put in good position
            ans[pos] = num
          
        
        return len(ans)


              

        

        
